### DISCUSSION TITLE
Column Selection

### DISCUSSION THREAD

**Discussion by wgetJane** [godotengine/godot-proposals#7004] [Editor]
currently, the only way to place multiple cursors in godot's editor is to hold alt then click individually for each cursor that you want, which is painfully slow

pretty much every other code editor i've worked with allows you to place many cursors very rapidly along a column, which is useful since a lot of the time i'd have a scenario like this:
```lua
example1.a = 1
example1.b = 2
example1.c = 3
example1.d = 4
example1.e = 5
```
if you wanted to change `example1` into `example2`, then with column selection you can simply create a column of cursors at the end of the variable name, hit backspace to erase the 1, then enter 2
alternatively, you can select the column of 1's and enter 2 to swiftly overwrite them

with godot's editor however, there doesn't seem to be a way to do this very quickly
(and using find and replace is overkill and still too slow)

other editors typically achieve this by letting you hold a key, usually shift+alt, then adding a new cursor to the line below or above using the arrow keys
(currently, using the up and down arrow keys with shift+alt doesn't do anything in godot's editor)

they also typically let you achieve the same thing but with the mouse
usually it's the ctrl or alt key, and while that key is held, you can click to start at a certain line and column
then drag vertically to create new cursors at the same column on the lines you move the mouse pointer to
or drag horizontally to select columns
(ctrl + click does already do something else in godot's editor, and so does alt + click, i guess shift+alt would work?)

this feature seems to be called "column select" or "box select" by people, but it effectively just works by automatically placing new cursors for you

here's how it looks in other editors:
https://youtube.com/watch?v=a6UN06Qk_Js
https://youtube.com/watch?v=DL0HB_qlSVk&t=16s
https://youtube.com/watch?v=6T6B3c3dPFI

**Comment by AThousandShips** [comment]
It's `Ctrl-Shift-Down/Up`, can be a bit confusing when it's another key for selecting, but it's in the editor action list

In the future please first turn to the other [community channels](https://godotengine.org/community)

**Reply by wgetJane** [reply]
damn, i tried a lot of possible shortcuts and i didn't think to try that one

still, creating a lot cursors really fast with the mouse isn't possible in godot's editor while it's possible in every other code editor

the keyboard method is fast for something like 3-4 lines, but using the mouse tends to be faster for more than that

> In the future please first turn to the other [community channels](https://godotengine.org/community)

i did ask there first but i was told that it's currently not a thing in godot, and that multi-caret support is still very recent

**Reply by AThousandShips** [reply]
They're a bit out of date 🙃 as the shortcuts were added back in October

**Comment by AThousandShips** [comment]
Block selection mode would be interesting, but it'd involve a lot of alterations and new details as it'd involve simulating selection areas where none exist etc.

**Reply by wgetJane** [reply]
idk if i'm oversimplying it, but wouldn't it just be making a bunch of cursors on the same column and then selecting text?

like if i used box/block selection on the column of 1's in the example with the mouse by holding the hypothetical box select key, then clicking to the left of the topmost one, then dragging to the right of the bottommost 1

isn't that achievable with the current multi-cursor system (albeit slower)? by moving the current single cursor to the left of the first one, then holding ctrl+shift, then pressing the down arrow key 4 times, then letting go off of ctrl+shift, then holding shift, then pressing the right arrow key

the only difference in other editors i think would be that you automatically revert back to one single cursor if you press the up or down arrow key after your box selection (current behaviour in godot is keeping all the cursors and just moving all of them them up or down)

**Reply by AThousandShips** [reply]
What happens when you move up to a line that has fewer columns? Also should it just be "add a caret on each line" or "add them in a box shape" or "select a box region by selecting `column a through b` on `lines c through d`?

To be clear I'm not dismissing this or pointing out reasons not to, this is for feature proposals and I'm bringing up things that need to be considered and solved and decided

**Reply by wgetJane** [reply]
> What happens when you move up to a line that has fewer columns?

sorry not sure what you mean exactly so i have 2 interpretations

if you mean what happens if you move the cursors after selection, then you can't, in other editors i've used they just deselect and revert back to normal when you press the up or down arrow key (you're back to 1 single cursor)

if you mean what happens if there's shorter lines during selection, i've seen different implementations

one is it creates "fake spaces" (idk what to call them) that turn into real spaces if you start typing

other solution is it just places the cursor at the end of the line

the latter is definitely the simplest way to do it, and i don't see any issues with doing that (in fact i'd say the solution that creates new spaces is kinda weird and unnecessary)
it wouldn't exactly be a "box" anymore of course, but you wouldn't be doing box selects with lines like that anyway

> Also should it just be "add a caret on each line" or "add them in a box shape" or "select a box region by selecting `column a through b` on `lines c through d`?

idk exactly what you mean here too, i guess i didn't explain it well, it's kinda difficult without just being there in person and showing my mouse movements to you

with the action i've described in the previous comment (holding the hypothetical key, then drag-clicking), what you would end up with is this selection and these cursors:
![image](https://github.com/godotengine/godot-proposals/assets/52103358/e8455541-5ed8-478e-ada9-529ddd5ea74a)
(i've achieved this in the godot editor using the specific set of keystrokes i also described in my previous comment)

**Reply by AThousandShips** [reply]
You got it pretty well yes, box selection as in "select column 2 and 3 on lines 4 and 5"

**Reply by wgetJane** [reply]
> if you mean what happens if there's shorter lines during selection, i've seen different implementations
> 
> one is it creates "fake spaces" (idk what to call them) that turn into real spaces if you start typing
> 
> other solution is it just places the cursor at the end of the line
> 
> the latter is definitely the simplest way to do it, and i don't see any issues with doing that (in fact i'd say the solution that creates new spaces is kinda weird and unnecessary) it wouldn't exactly be a "box" anymore of course, but you wouldn't be doing box selects with lines like that anyway

actually another simple solution is to *not* place cursors on shorter lines

which i think is actually the best way to do it, since it would make it easier to work with block selects for something like this:

```lua
example1.a = 1

example1.b = 2

example1.c = 3

example1.d = 4

example1.e = 5
```

**Comment by voylin** [comment]
Kind of hope this gets implemented at some point ^^
